HSEARCH-4305 + HSEARCH-4708 No automatic reindexing when setting a property annotated with @OneToOne(mappedBy = ...) @IndexedEmbedded to null

yrodiere opened the pull request:
* [HSEARCH-4708](https://hibernate.atlassian.net/browse/HSEARCH-4708): No automatic reindexing when changing a property annotated with @OneToOne(mappedBy = ...) @IndexedEmbedded
* [HSEARCH-4305](https://hibernate.atlassian.net/browse/HSEARCH-4305): No automatic reindexing when setting a property annotated with @OneToOne(mappedBy = ...) @IndexedEmbedded to null

This bug turned out to be quite complex to fix, since I essentially had to add a brand new feature to Hibernate Search: when an association is marked as dirty, Hibernate Search will automatically retrieve the old and new targets of the association, and mark the inverse side of that association as dirty on those entities.

This is limited to the owning side of `ToOne` associations at the moment (see below why), but we will probably expand that to all associations at some point ([HSEARCH-3567](https://hibernate.atlassian.net/browse/HSEARCH-3567)).

In any case, this circumvents the problems of not getting any event for changes on the non-owning side of `@OneToOne` associations: we expect both sides to be updated, so when the owning side of the association changes, this feature will automatically mark the non-owning side as dirty.

I'm going to backport this to 6.1 once we've checked this works fine everywhere (I'm especially worried about ORM 6, I'll run some tests locally).

This PR is best reviewed commit by commit, because there's a lot going on here. I'd recommend ignoring the commits that just add new tests, because those are huge (sorry, but there's a lot of different cases to test).

yrodiere commented:
So, there are still a few failures... I'm working on it.

yrodiere commented:
@marko-bekhta This should be finally ready for review... sorry for the delay.

marko-bekhta commented:
Hey, no worries 😃

yrodiere commented:
@marko-bekhta I pushed two fixup commits to address the build failures. Let's not merge just yet, I'll wait for your review then I'll address it and squash the commits as appropriate :)

marko-bekhta reviewed (commented):
This is a big one to grasp. From the parts that I've managed to figure out - all makes sense 😃 
 
I was also looking at `PojoPathDefinition` to see if it would make sense to have two variants of it (with and without state), but in the end, I've discarded the idea as it would most likely make the code more cumbersome.

Other than a few inline questions, there's not much more from me 😃

yrodiere commented on mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java:
```diff
@@ -128,39 +141,52 @@ public void onPostUpdate(PostUpdateEvent event) {
+		BitSet dirtyDirectAssociationPaths;
 		if ( dirtyCheckingEnabled ) {
 			int[] dirtyProperties = event.getDirtyProperties();
 			if ( dirtyProperties == null || dirtyProperties.length == 0 ) {
```
It might, but that's also an unlikely scenario. *Absolutely* all properties being changed, all at once, in the same transaction... I can't see that being frequent save for a few exotic applications.
Feel free to create a ticket if you think there's something to be explored, though.

yrodiere commented on orm6/integrationtest/mapper/orm/ant-src-changes.patch:
```diff
@@ -1,8 +1,377 @@
+diff --git a/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/association/bytype/AbstractAutomaticIndexingAssociationBaseIT.java b/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/association/bytype/AbstractAutomaticIndexingAssociationBaseIT.java
+index cbd1ace7ae..193d70675f 100644
```
I rebase `wip/main/dependency-update/orm6-in-main-code` on my branch, then generate patches using `orm6/extract-patches.sh`. See https://github.com/hibernate/hibernate-search/blob/3a339d10300b75068a71580bcd65381bba9d2185/orm6/README.md

And before you say it: I agree it's agree and impractical, it's just the least bad solution I found. I can't wait to get rid of it in Hibernate Search 7, when we upgrade to ORM 6 :)

yrodiere commented on mapper/orm/src/main/java/org/hibernate/search/mapper/orm/logging/impl/Log.java:
```diff
@@ -305,4 +306,14 @@ SearchException unknownClassForMappedEntityType(@FormatWith(ClassFormatter.class
+					+ " The exception is being ignored to preserve backwards compatibility with earlier versions of Hibernate Search."
+					+ " Failure: %3$s"
+					+ " %2$s") // Context
+	void failedToResolveStateRepresentation(String path, EventContext context, String causeMessage,
```
Absolutely, good catch.

yrodiere commented:
Thanks! I addressed your comments as well as a few code smells detected by Sonar, and rebased. Let's wait for CI and merge.

> This is a big one to grasp. From the parts that I've managed to figure out - all makes sense smiley

Right, perhaps not the best PR for your first review, but I guess if you survived this, you came out stronger :grin: 

> I was also looking at PojoPathDefinition to see if it would make sense to have two variants of it (with and without state), but in the end, I've discarded the idea as it would most likely make the code more cumbersome.

Right, two variants is probably not the best, especially if we need to add more optional content later.
